### ISSUE TITLE
support syntax highlighting for fenced code block

### ISSUE THREAD

**Issue by jubalfh** [glitch-soc/mastodon#2011]
### Pitch

allowing syntax highlighting for fenced code blocks would greatly improve the readability of the code, and if paired with ability to mark specific lines or blocks of lines, it would enhance comments on the code snippets significantly.

### Motivation

glitch-soc already supports fenced code blocks – this would offer just a visual enhancement of the feature; the target audience is the same as the users of markdown and fenced code blocks.

**Comment by Plastikmensch** [comment]
I support this in theory, but will be a huge undertaking and I'm not sure it's worth the work required for this.
You would have to first create a list of supported code languages, define the syntax for each, run RegEx over the code blocks and add the styling to it.
It makes sense for GitHub to do, but I don't think it makes sense for Glitch.

**Comment by jubalfh** [comment]
ah, but this is a solved problem for a very long time: see for example [highlight.js](https://highlightjs.org/)

**Comment by Plastikmensch** [comment]
Okay, I played around a bit:
Language detection is wonky at best.
The js test code I used got either detected as arcade or awk.
I would let users specify the language, similar to how github does it.
This needs modification of markdown processing as that is being stripped.
It works* by making a few adjustments to `status_content.js`, running highlight on `contentHtml`.
It's very error prone though, as this needs careful string manipulation to get anything between html tags inside code blocks.

**Comment by Plastikmensch** [comment]
To illustrate the problem:
Specifying js:
![Screenshot_2022-12-15_20-28-57](https://user-images.githubusercontent.com/29483052/207949888-ac51696a-ccf8-4ae9-8c73-7e8288519383.png)


Not specifying js:
![Screenshot_2022-12-15_20-25-42](https://user-images.githubusercontent.com/29483052/207949597-6df71b53-4e60-4e08-a7a9-53e68824c64a.png)

**Comment by jubalfh** [comment]
i guess that to be on the safe side, the highlighter should not be run unless the language is explicitly specified – or it should be run with the plaintext format explicitly.

**Comment by Plastikmensch** [comment]
> i guess that to be on the safe side, the highlighter should not be run unless the language is explicitly specified – or it should be run with the plaintext format explicitly.

Agreed.
Again, this requires modification to the Markdown processor, because anything immediately after ``` gets stripped, and it needs to be passed to the frontend, so it can be used.
It should not run at all if no language is specified imo though.

**Comment by ionathanch** [comment]
I'd really like this, what's the status on the draft PR?

**Comment by Plastikmensch** [comment]
> I'd really like this, what's the status on the draft PR?

It's working fine, but I don't like how much of a mess it became and would rather redo it.
The main issue was that glitch-soc doesn't add a background to code blocks, which messes up highlighting in the reply preview and because of that any language value was available in the preview. The latter was fixed by only allowing valid languages in the backend and remove it there before it's passed to the frontend.

TL;DR: I'm not happy with it in it's current state.
